File: tests/output/scala/src/test/scala/algoliasearch/client/RecommendTest.scala
package algoliasearch.client

import algoliasearch.api.RecommendClient
import algoliasearch.config.*
import algoliasearch.{EchoInterceptor, assertError}
import algoliasearch.recommend.*
import algoliasearch.exception.*
import org.json4s.*
import org.json4s.native.JsonParser.*
import org.scalatest.funsuite.AnyFunSuite

import scala.concurrent.duration.Duration
import scala.concurrent.{Await, ExecutionContextExecutor}

class RecommendTest extends AnyFunSuite {
  implicit val ec: ExecutionContextExecutor = scala.concurrent.ExecutionContext.global
  implicit val formats: Formats = org.json4s.DefaultFormats

  def testClient(appId: String = "appId", apiKey: String = "apiKey"): (RecommendClient, EchoInterceptor) = {
    val echo = EchoInterceptor()
    (
      RecommendClient(
        appId = appId,
        apiKey = apiKey,
        clientOptions = ClientOptions
          .builder()
          .withRequesterConfig(requester => requester.withInterceptor(echo))
          .build()
      ),
      echo
    )
  }

  test("calls api with correct read host") {

    val (client, echo) = testClient(appId = "test-app-id", apiKey = "test-api-key")

    Await.ready(
      client.get[Any](
        path = "/test"
      ),
      Duration.Inf
    )

    assert(echo.lastResponse.get.host == "test-app-id-dsn.algolia.net")
  }

  test("calls api with correct write host") {

    val (client, echo) = testClient(appId = "test-app-id", apiKey = "test-api-key")

    Await.ready(
      client.post[Any](
        path = "/test"
      ),
      Duration.Inf
    )

    assert(echo.lastResponse.get.host == "test-app-id.algolia.net")
  }

  test("calls api with correct user agent") {
    val (client, echo) = testClient()

    Await.ready(
      client.post[Any](
        path = "/test"
      ),
      Duration.Inf
    )

    val regexp =
      """^Algolia for scala \(\d+\.\d+\.\d+(-.*)?\)(; [a-zA-Z. ]+ (\(\d+((\.\d+)?\.\d+)?(-.*)?\))?)*(; Recommend (\(\d+\.\d+\.\d+(-.*)?\)))(; [a-zA-Z. ]+ (\(\d+((\.\d+)?\.\d+)?(-.*)?\))?)*$""".r
    val header = echo.lastResponse.get.headers("user-agent")
    assert(header.matches(regexp.regex), s"Expected $header to match the following regex: ${regexp.regex}")
  }

  test("calls api with default read timeouts") {
    val (client, echo) = testClient()

    Await.ready(
      client.get[Any](
        path = "/test"
      ),
      Duration.Inf
    )

    assert(echo.lastResponse.get.connectTimeout == 2000)
    assert(echo.lastResponse.get.responseTimeout == 5000)
  }

  test("calls api with default write timeouts") {
    val (client, echo) = testClient()

    Await.ready(
      client.post[Any](
        path = "/test"
      ),
      Duration.Inf
    )

    assert(echo.lastResponse.get.connectTimeout == 2000)
    assert(echo.lastResponse.get.responseTimeout == 30000)
  }
}

File: tests/output/python/tests/client/recommend_test.py
from re import compile

from algoliasearch.http.transporter import EchoTransporter
from algoliasearch.recommend.client import RecommendClient
from algoliasearch.recommend.config import Config


class TestRecommendClient:
    _config: Config
    _client: RecommendClient

    def create_client(self) -> RecommendClient:
        self._config = Config("appId", "apiKey")
        self._client = RecommendClient(EchoTransporter(self._config), self._config)

    async def test_api_0(self):
        self._client = RecommendClient(
            EchoTransporter(Config("test-app-id", "test-api-key")),
            Config("test-app-id", "test-api-key"),
        )

        _req = await self._client.custom_get_with_http_info(
            path="/test",
        )

        assert _req.host == "test-app-id-dsn.algolia.net"

    async def test_api_1(self):
        self._client = RecommendClient(
            EchoTransporter(Config("test-app-id", "test-api-key")),
            Config("test-app-id", "test-api-key"),
        )

        _req = await self._client.custom_post_with_http_info(
            path="/test",
        )

        assert _req.host == "test-app-id.algolia.net"

    async def test_common_api_0(self):
        self.create_client()

        _req = await self._client.custom_post_with_http_info(
            path="/test",
        )

        regex_user_agent = compile(
            "^Algolia for Python \\(\\d+\\.\\d+\\.\\d+(-?.*)?\\)(; [a-zA-Z. ]+ (\\(\\d+((\\.\\d+)?\\.\\d+)?(-?.*)?\\))?)*(; Recommend (\\(\\d+\\.\\d+\\.\\d+(-?.*)?\\)))(; [a-zA-Z. ]+ (\\(\\d+((\\.\\d+)?\\.\\d+)?(-?.*)?\\))?)*$"
        )
        assert regex_user_agent.match(_req.headers.get("user-agent")) is not None

    async def test_common_api_1(self):
        self.create_client()

        _req = await self._client.custom_get_with_http_info(
            path="/test",
        )

        assert _req.timeouts.get("connect") == 2000
        assert _req.timeouts.get("response") == 5000

    async def test_common_api_2(self):
        self.create_client()

        _req = await self._client.custom_post_with_http_info(
            path="/test",
        )

        assert _req.timeouts.get("connect") == 2000
        assert _req.timeouts.get("response") == 30000